### PR TITLE
Feat/coredo model attrs

### LINKED ISSUE #896: Implement converter between model and coredoc.
#### Testing
unit tests to cover conversion

#### Acceptance Criteria
conversion works

### PR DESCRIPTION
**Pull request by vedhavyas**
Closes #896

### CODE REVIEW

File: documents/model.go [hunk 1]
```diff
@@ -118,14 +118,17 @@ type Model interface {
 	IsDIDCollaborator(did identity.DID) (bool, error)
 
 	// AddAttribute adds a custom attribute to the model with the given value. If an attribute with the given name already exists, it's updated.
-	AddAttribute(keyLabel string, attributeType attributeType, value string) (*CoreDocument, error)
+	AddAttribute(attr Attribute) (*CoreDocument, error)
```
[the remark this answers is not in the document]
Comment (manuelpolzhofer) [reply documents/model.go]: As discussed. Returning the model in that interface and the CoreDocument only in coredocument interface

File: documents/converters.go [hunk 2]
```diff
@@ -433,3 +434,105 @@ func convertNFTs(tokenRegistry TokenRegistry, nfts []*coredocumentpb.NFT) (nnfts
 	}
 	return nnfts, err
 }
+
+// toP2PAttributes convert model attributes to p2p attributes
+// since the p2p respresentation of attributes is a list, we will always sort the keys and then insert to the list.
+func toP2PAttributes(attrs map[AttrKey]Attribute) (pattrs []*coredocumentpb.Attribute, err error) {
```
Comment (mikiquantum) [inline documents/converters.go]:
```suggestion
func toProtocolAttributes(attrs map[AttrKey]Attribute) (pattrs []*coredocumentpb.Attribute, err error) {
```
I would like to use the word `protocol` here, as it is a protocol format that is used in P2P communications.

File: utils/byteutils/bytes.go [hunk 3]
```diff
@@ -77,3 +79,33 @@ func IsBitSet(n byte, pos uint) bool {
...
+		return false
+	}
+}
+
+// Swap swaps the i, j values with in the array.
+func (a BytesArray) Swap(i, j int) {
+	a[j], a[i] = a[i], a[j]
+}
+
+// SortByte32Slice sorts the byte32 slices
```
Comment (mikiquantum) [inline utils/byteutils/bytes.go]: Can you add an explicit comment on how the slice is sorted? ascendent, descendent... Easier to reuse later.

File: documents/converters.go [hunk 4]
```diff
@@ -433,3 +434,105 @@ func convertNFTs(tokenRegistry TokenRegistry, nfts []*coredocumentpb.NFT) (nnfts
...
+			if err != nil {
+				return nil, err
+			}
+			pattr.Value = &coredocumentpb.Attribute_ByteVal{ByteVal: b}
+		case AttrString:
+			pattr.Value = &coredocumentpb.Attribute_StrVal{StrVal: attr.Value.Str}
+		case AttrBytes:
+			pattr.Value = &coredocumentpb.Attribute_ByteVal{ByteVal: attr.Value.Bytes}
+		case AttrTimestamp:
+			pattr.Value = &coredocumentpb.Attribute_TimeVal{TimeVal: attr.Value.Timestamp}
```
Comment (mikiquantum) [inline documents/converters.go]: Should we add `default` and error out?

File: documents/converters.go [hunk 5]
```diff
@@ -433,3 +434,105 @@ func convertNFTs(tokenRegistry TokenRegistry, nfts []*coredocumentpb.NFT) (nnfts
...
+		}
+
+		pattrs = append(pattrs, pattr)
+	}
+
+	return pattrs, nil
+}
+
+const attributeP2PPrefix = "ATTRIBUTE_TYPE_"
```
Comment (mikiquantum) [inline documents/converters.go]:
```suggestion
const attributeTypeProtocolPrefix = "ATTRIBUTE_TYPE_"
```

File: documents/coredocument.go [hunk 6]
```diff
@@ -217,26 +219,30 @@ func (cd *CoreDocument) PrepareNewVersion(documentPrefix []byte, collaborators C
...
-// copyAttrMap copies the attributes map.
-func copyAttrMap(attributes map[AttrKey]Attribute) map[AttrKey]Attribute {
-	m := make(map[AttrKey]Attribute)
-	for k, v := range attributes {
-		m[k] = Attribute{
-			KeyLabel: v.KeyLabel,
-			Key:      v.Key,
-			Value:    v.Value,
-		}
+// updateAttributes updates the p2p attributes with new one and returns the both the formats
```
Comment (mikiquantum) [inline documents/coredocument.go]:
```suggestion
// updateAttributes updates the p2p attributes with new ones and returns both formats
```

File: documents/converters.go (hunk 4 above)
Comment (vedhavyas) [reply documents/converters.go]: not needed since it checked above

**Pull request merged**